plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/UIUtils.java

```diff
@@ -1803,6 +1807,94 @@ public static <T> T syncExec(RunnableWithResult<T> runnable) {
             return null;
         }
     }
+    
+    private static long getLongOperationTime() {
+        try {
+            return PlatformUI.getWorkbench().getProgressService().getLongOperationTime();
+        } catch (Exception ex) { // when workbench is not initialized yet during startup
+            return 800; // see org.eclipse.ui.internal.progress.ProgressManager.getLongOperationTime()
+        }
+    }
+    
+    /**
+     * Execute runnable task synchronously while displaying job indeterminate indicator and blocking the UI, when called from the UI thread
+     */
+    @NotNull
+    public static <T> T syncExecBlocking(
+        @NotNull String operationDesc,
+        @NotNull RunnableWithResult<T> runnable
+    ) throws InvocationTargetException, InterruptedException {
+        
+        @SuppressWarnings("unchecked")
+        RunnableWithResultEx<T> runnableEx = runnable instanceof RunnableWithResultEx ? (RunnableWithResultEx<T>) runnable : null;
+        
+        final AbstractJob job = new AbstractJob(operationDesc) {
+            @Override
+            protected IStatus run(DBRProgressMonitor monitor) {
+                monitor.beginTask(operationDesc, IProgressMonitor.UNKNOWN);
+                runnable.run();
+                monitor.done();
+                return Status.OK_STATUS;
+            }
+            
+            @Override
+            protected void canceling() {
+                if (runnableEx != null) {
+                    runnableEx.cancel();
+                }
+            }
+        };
+        job.schedule();
+        
+        if (UIUtils.isUIThread()) {
+            Display display = getDisplay();
+            if (!display.isDisposed()) {
+                RunnableWithResultEx<Boolean> shortWait = new RunnableWithResultEx<Boolean>() {
+                    @Override
+                    protected Boolean runWithResultImpl() throws Exception {
+                        return !job.join(getLongOperationTime(), new NullProgressMonitor());
+                    }
+
+                    @Override
+                    protected void onCancelled() { }
+                };
+                Collection<Shell> shells = Arrays.asList(display.getShells());
+                shells.forEach(s -> s.setEnabled(false));
```

Comment:
Uh, let's use plain `for each` loop here along with removing unnecessary conversion into a collection